src/util/DeferredProperty.java

```diff
@@ -0,0 +1,231 @@
+     * @return True if this instance is dirty.
+     */
+    public boolean isDirty() {
+        return isDirty;
+    }
+
+    /**
+     * Used to check if this instance is _not_ marked as dirty.
+     *
+     * @return True if this instance is not dirty.
+     */
+    public boolean isClean() {
+        return !isDirty();
+    }
+
+    /**
+     * Set the {@link DeferredProperty#value} and cleans this instance by not marking it as dirty.
+     * Calling this function when this instance is not dirty will just set {@link DeferredProperty#value}.
+     *
+     * @param value The value to set this property {@link DeferredProperty#value} to.
+     * @return Returns <code>true</code> if the instance was cleaned by marking it as no longer dirty.
+     */
+    public boolean set(T value) {
+        this.value = value;
+        return clean();
+    }
+
+    /**
+     * Sets the {@link DeferredProperty#value} but only if this is dirty.
+     * Calling this function when this instance is not dirty will not set the value.
+     * However, the current value of this property will always be returned.
+     *
+     * @param value The value to set this property {@link DeferredProperty#value} to.
+     * @return Returns the value of this instance value even if no new assignment was made.
+     */
+    public T trySet(T value) {
+        if (isDirty()) {
+            set(value);
+        }
+        return get();
+    }
+
+    /**
+     * Sets the {@link DeferredProperty#value} but only if this is dirty.
+     * The new {@link DeferredProperty#value} would be the one returned by the <code>supplier</code>.
+     * Calling this function when this instance is not dirty will not set the value.
+     * However, the current value of this property will always be returned.
+     *
+     * @param supplier The supplier that returns the new value of {@link DeferredProperty#value}.
+     * @return Returns the value of this instance value even if no new assignment was made.
+     */
+    public T trySet(Supplier<T> supplier) {
+        return trySet(supplier.get());
+    }
+
+    /**
+     * If a supplier has be assigned and this is dirty then a new value is calcalted and returned.
+     * {@link Optional<T>} is used to highlight that if this instance is dirty then {@link Optional#empty()} is returned instead of <code>null</code>.
+     *
+     * @return If dirty then {@link Optional#empty()} else {@link Optional#of(Object)} with {@link DeferredProperty#value}.
+     */
+    public Optional<T> tryCompute() {
+        // Automatically update the value of this property if dirty.
+        if (isDirty() && supplier != null) {
+            set(supplier.get());
+        }
+
+        return isDirty() ? Optional.empty() : Optional.of(value);
+    }
+
+    /**
+     * Calls {@link DeferredProperty#tryCompute()} and returns null if the {@link Optional#isEmpty()}.
+     *
+     * @return Returns <code>null</code> if this is dirty and no supplier exists.
+     */
+    public T get() {
```

Comment:
```suggestion
    public T getValue() {
```

I guess in normal use people should call get() for getting the value that takes isDirty() into account. But the function name get() does not specify what it is getting. So I suggest to change get() to getValue() and getValue() to getRawValue() (or maybe getUncheckedValue()).